### PR TITLE
Adding new PyK wrappers to LLVM Python Bindings

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
This PR exposes the new infrastructure created by the LLVm Backend to parse the Hints file.
We introduce the wrappers to make the streaming parser possible to use as a list of events and access it through an integrator. Or not loading all the events at the same time but accessing them one per time through the `next()` function.

Depends on: https://github.com/runtimeverification/llvm-backend/pull/1090

### CODE REVIEW

**Review by theo25 — APPROVED** [review]
This looks good to me, but let’s wait for at least one approval from the RV K team before merging.

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (Baltoli) [inline pyk/src/pyk/kllvm/hints/prooftrace.py]:
```suggestion
        """Returns the version of the binary hints format used by this trace."""
```

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (Baltoli) [inline pyk/src/pyk/kllvm/hints/prooftrace.py]: These documentation comments are nice to have, but are really just re-stating the information in the type signatures of the methods. Consider something more like:
```suggestion
        """Returns a list of events that occurred before the initial configuration was constructed."""
```
...and so on for the other instances.

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (Baltoli) [inline pyk/src/pyk/kllvm/hints/prooftrace.py]: Ideally let's keep documentation to the same line length as the actual code

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (Baltoli) [inline pyk/src/pyk/kllvm/hints/prooftrace.py:388]: Are these comments something that Sphinx will understand?

**Review by Baltoli — DISMISSED** [review]
A couple of minor comments, and I'll let @tothtamas28 provide a final review as well in case I've missed anything.

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (tothtamas28) [reply pyk/src/pyk/kllvm/hints/prooftrace.py:388]: Good point, please use the Sphinx docstring format: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

An example: https://kframework.org/pyk/_modules/pyk/ktool/krun.html#llvm_interpret

How it renders: https://kframework.org/pyk/api/pyk.ktool.krun.html#pyk.ktool.krun.llvm_interpret

Comment (Robertorosmaninho) [reply pyk/src/pyk/kllvm/hints/prooftrace.py:388]: Alright! I was using the [Google Style Python Docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) pattern because I coulnd't find one in the kllvm module, but I can switch to the `sphinx` style as well! 

The first style was rendered as shown [here](https://kframework.org/pyk/api/pyk.kllvm.hints.prooftrace.html#module-pyk.kllvm.hints.prooftrace)!

Comment (tothtamas28) [reply pyk/src/pyk/kllvm/hints/prooftrace.py:388]: I'm not against adopting this format if there's support for it in Sphinx. We do not have a large amount of docstrings yet so it's not late to make this the format used in pyk if there's anyone in favor.

Should we discuss this during the K dev meeting?

Comment (tothtamas28) [reply pyk/src/pyk/kllvm/hints/prooftrace.py:388]: > The first style was rendered as shown [here](https://kframework.org/pyk/api/pyk.kllvm.hints.prooftrace.html#module-pyk.kllvm.hints.prooftrace)!

This looks like the docstring text without any styling applied. But the plugin you linked

> [Google Style Python Docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

can handle this format properly I think.

Comment (Robertorosmaninho) [reply pyk/src/pyk/kllvm/hints/prooftrace.py:388]: I think we can definitely discuss this in the K dev meeting. I honestly prefer this style, as it seems easier to read and write in the code. I missed this extension, though: https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html which is responsible for converting the style and correctly generating the format we expect to see after rendering it!

Comment (tothtamas28) [reply pyk/src/pyk/kllvm/hints/prooftrace.py:388]: Opened an issue:
* #4435

**Comment by Robertorosmaninho** [comment]
This is the new rendered comments: 
image

File: pyk/docs/conf.py
Comment (tothtamas28) [inline pyk/docs/conf.py]:
```suggestion
    'sphinx.ext.napoleon',
    'sphinx.ext.viewcode',
```

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (tothtamas28) [inline pyk/src/pyk/kllvm/hints/prooftrace.py]:
```suggestion
        """Return the axiom ordinal number of the rewrite rule.
        
        The rule ordinal represents the `nth` axiom in the kore definition.
        """
```

https://peps.python.org/pep-0257/

File: pyk/src/pyk/kllvm/hints/prooftrace.py
Comment (tothtamas28) [inline pyk/src/pyk/kllvm/hints/prooftrace.py]:
```suggestion
            A string representation of the LLVMRuleEvent object using the AST printing method.
```

The type hints already carry this information.

File: pyk/src/pyk/testing/_kompiler.py
Comment (tothtamas28) [inline pyk/src/pyk/testing/_kompiler.py:395]: This looks a bit hacky. Consider implementing a dedicated cache of hints inputs for `ProofTraceTest`. Then no renaming is needed, and `kompiler` can be dropped as a direct dependency.

(After the renaming, the cache might no longer point to the right file.)

File: pyk/src/pyk/testing/_kompiler.py
Comment (tothtamas28) [inline pyk/src/pyk/testing/_kompiler.py]: Paths support this:

```suggestion
        path = definition_dir / header_file_name
        with path.open('wb') as f:
```

**Review by tothtamas28 — APPROVED** [review]

File: pyk/src/pyk/testing/_kompiler.py
Comment (Robertorosmaninho) [reply pyk/src/pyk/testing/_kompiler.py:395]: The renaming is just for creating a new file with a custom name instead of simply ‘hints.bin’ or something static like this. This is mainly to avoid the use of 2 different tests appending in the same file (an issue that I found while writing those tests).

**Comment by Robertorosmaninho** [comment]
Ping @Baltoli

**Review by Baltoli — APPROVED** [review]

**Pull request merged**
